api/base/views.py

```diff
@@ -437,7 +453,71 @@ def root(request, format=None, **kwargs):
     if utils.has_admin_scope(request):
         return_val['meta']['admin'] = True
 
-    return Response(return_val)
+    resp = Response(return_val)
+
+    for key, value in cookies.items():
+        resp.set_cookie(key, value)
+
+    return resp
+
+
+def sloan_study_disambiguation(request):
+    """
+    This is a hack to set flags and cookies for out Sloan study, it can be deleted when the study is complete.
+    """
+    user = request.user
+    cookies = {}
+
+    flags = []
+    sloan_data = {}
+    for flag in Flag.objects.all():
+        active = flag.is_active(request._request)
+        if flag.name in SLOAN_FLAGS and flag.everyone is None:
+            sloan_data[flag.name] = active
+
+        # User tags should override any cookie info
+        if user and not user.is_anonymous:
+            if user.all_tags.filter(name=flag.name):
+                sloan_data[flag.name] = True
+                active = True
+            if user.all_tags.filter(name=f'no_{flag.name}'):
+                sloan_data[flag.name] = False
+                active = False
+
+        if active:
+            flags.append(flag.name)
+
+    # This is used exclusively for our partnership with Sloan and can be deleted after their study is complete.
+    referer_url = request.environ.get('HTTP_REFERER', '')
+    provider_domains = list(PreprintProvider.objects.exclude(domain='').values_list('domain', flat=True))
+
+    provider_domains = [domains for i, domains in enumerate(provider_domains) if referer_url.startswith(domains)]
+
+    provider = None
+    if provider_domains:
+        provider = PreprintProvider.objects.get(domain=provider_domains[0])
+
+    if referer_url.startswith(DOMAIN + 'preprints/'):
+        provider = PreprintProvider.objects.get(_id=referer_url.split('/')[4])
```

Comment:
this 500s for the OSF preprints landing page (`https://<domain>/preprints/`), which seems like a good test case to add

i generally consider any sort of `foo.split(...)[n]` an antipattern -- probably cleaner to extract the provider with a simple regex (and use the same one instead of the `startswith` check above)

Reply:
Yes this is bad and regex is probably better, but there's no obvious reverse/resolve function I can find that handles custom domains. I'm a little perplexed by how the front end is handling the routing here, but I'm really hampered because I can't reverse the url automatically to see what page is visible.